flink-state-backends/flink-statebackend-rocksdb/src/main/java/org/apache/flink/contrib/streaming/state/RocksDBKeyedStateBackendBuilder.java

```diff
@@ -0,0 +1,478 @@
+			// restore operations may reconfigure the key serializer, so accessing the key serializer
+			// only now we can be certain that the key serializer used in the builder is final.
+			sharedRocksKeyBuilder = new RocksDBSerializedCompositeKeyBuilder<>(
+				keySerializerProvider.currentSchemaSerializer(),
+				keyGroupPrefixBytes,
+				32);
+			// init snapshot strategy after db is assured to be initialized
+			snapshotStrategy = initializeSavepointAndCheckpointStrategies(cancelStreamRegistry, rocksDBResourceGuard,
+				kvStateInformation, keyGroupPrefixBytes, db, backendUID, materializedSstFiles, lastCompletedCheckpointId);
+			// init priority queue factory
+			priorityQueueFactory = initPriorityQueueFactory(
+				keyGroupPrefixBytes, kvStateInformation, db, writeBatchWrapper, nativeMetricMonitor);
+		} catch (Throwable e) {
+			// Do clean up
+			IOUtils.closeQuietly(cancelStreamRegistry);
+			IOUtils.closeQuietly(writeBatchWrapper);
+			IOUtils.closeQuietly(defaultColumnFamilyHandle);
+			IOUtils.closeQuietly(nativeMetricMonitor);
+			IOUtils.closeAllQuietly(kvStateInformation.values());
+			IOUtils.closeQuietly(db);
+			// it's possible that db has been initialized but later restore steps failed
+			IOUtils.closeQuietly(restoreOperation);
+			IOUtils.closeQuietly(columnFamilyOptions);
+			IOUtils.closeQuietly(dbOptions);
+			IOUtils.closeQuietly(writeOptions);
+			try {
+				FileUtils.deleteDirectory(instanceBasePath);
+			} catch (Exception ex) {
+				LOG.warn("Failed to instance base path for RocksDB: " + instanceBasePath, ex);
+			}
+			// Log and rethrow
+			if (e instanceof BackendBuildingException) {
+				throw (BackendBuildingException) e;
+			} else {
+				String errMsg = "Caught unexpected exception.";
+				LOG.error(errMsg, e);
+				throw new BackendBuildingException(errMsg, e);
+			}
+		}
+		return new RocksDBKeyedStateBackend<>(
+			this.operatorIdentifier,
+			this.userCodeClassLoader,
+			this.instanceBasePath,
+			this.dbOptions,
+			this.columnFamilyOptions,
+			this.kvStateRegistry,
+			this.keySerializerProvider,
+			this.numberOfKeyGroups,
+			this.keyGroupRange,
+			this.executionConfig,
+			this.numberOfTransferingThreads,
+			this.ttlTimeProvider,
+			db,
+			kvStateInformation,
+			keyGroupPrefixBytes,
+			cancelStreamRegistry,
+			this.keyGroupCompressionDecorator,
+			rocksDBResourceGuard,
+			snapshotStrategy.checkpointSnapshotStrategy,
+			snapshotStrategy.savepointSnapshotStrategy,
+			writeBatchWrapper,
+			defaultColumnFamilyHandle,
+			nativeMetricMonitor,
+			sharedRocksKeyBuilder,
+			priorityQueueFactory
+		);
+	}
+
+	private AbstractRocksDBRestoreOperation<K> getRocksDBRestoreOperation(int keyGroupPrefixBytes,
+																CloseableRegistry cancelStreamRegistry,
+																LinkedHashMap<String, StateColumnFamilyHandle> kvStateInformation,
+																RocksDBNativeMetricMonitor nativeMetricMonitor) {
+		if (restoreStateHandles == null || restoreStateHandles.isEmpty()) {
```

Comment:
This `null` check is not required and can be confusing now, because by contract `restoreStateHandles` is expected to be `@Nonnull`